### PR TITLE
config file: tolerate trailing whitespaces and comments. Fix #3201

### LINKED ISSUE #3201: Config file should tolerate pending whitespace and comments
The following configs cause lightningd to fail to launch:

```
# Trailing space after 5000
fee-base=5000 
```

```
fee-base=5000 # 5 satoshi
```

### PR DESCRIPTION
**Pull request by gorazdko**
## Status
WIP

### CODE REVIEW

File: lightningd/options.c [hunk 1]
```diff
@@ -759,6 +759,23 @@ static void opt_parse_from_config(struct lightningd *ld, bool early)
 		else {
 			/* Only valid forms are "foo" and "foo=bar" */
 			all_args[i] = tal_fmt(all_args, "--%s", lines[i]);
+
+			char *pos = strchr(all_args[i], '#');
```
Comment (arowser) [inline lightningd/options.c]: Error in follow config file:
```
   bitcoin-rpcpassword=xxx#pass  # xxx
```
Maybe the fix should add to ccan/ccan/opt/parse.c

Comment (darosior) [reply lightningd/options.c]: ccan fixes should be integrated into ccan directly https://github.com/rustyrussell/ccan/

Comment (gorazdko) [reply lightningd/options.c]: tnx for suggestions, i'll find a cleaner approach.

Core doesnt allow such passwords, so neither should we:

> bitcoin-rpcpassword=xxx#pass  # xxx

(16115)gorazd@gorazd-MS-7C37:~/Projects/bitcoin$ ./src/bitcoind
Error: Error reading configuration file: parse error on line 5, using # in rpcpassword can be ambiguous and should be avoided

Comment (gorazdko) [reply lightningd/options.c]: tnx ill find a cleaner approach

i see now how Core doesnt allow neither # in passwords nor trailing comments after passwords:

> Error: Error reading configuration file: parse error on line 6, using # in rpcpassword can be ambiguous and should be avoided

**Comment by rustyrussell** [comment]
If you start down this path, how do you use `#` or ` ` in a name, etc?

No, I think we have to stick with exact usage; we just need to make sure that the messages are clear if it's accidental.

**Comment by gorazdko** [comment]
I wanted to make the parser compatible with Core (remove trailing comments, initial and trailing white-spaces which this PR does), but yes, i see this brings more bad things because Core too has unhandled corner cases:

* they allow trailing comments everywhere except for the rpcpassword where they throw an error, everywhere else # is considered a comment and cannot be part of a name (no warnings thrown)
  * they allow trailing comments at torpassword without a warning though
* they trim the trailing spaces which could be a part of the password



* therefore the only thing that could be done here is to throw:
  * error message:    "config file line X: '#' can be used only at the beginning of the line to denote a comment"
  * error message:    "config file line X: line must not begin with a space "
  * warning message:  "config file line X: trailing spaces are not recommended"
  * also fix the line counting:

  ```
  fee-base=5000


  line4

  ```
  * lightning config file line 2: line4: unrecognized option


I can force push these changes here if they are reasonable, otherwise the issue can be closed.

**Comment by cdecker** [comment]
I'm afraid it's going to be a NACK from me as well. Starting with these exceptions is a never-ending rabbit hole.

**Pull request closed without merging**